# Vorlesung: Arbeiten mit Webseiten als Quellen

Viele Datenbank verfügen über eine Webseite, die eine Textsuche erlaubt. Das Result einer solchen Suche kann über das Python Paket beautifulsoup ausgelesen und in einen Dataframe geschrieben werden. 
Aufgrund dieser Daten kann dann weiter gearbeitet werden, sei es mit regulären Ausdrücken oder NLTK. 

In [5]:
import json
import re
import pandas as pd

import requests
from bs4 import BeautifulSoup

## Stellen einer Suchanfrage: Jordanus

Als erstes muss in der Suchmaske der Webseite eine Suche gestartet werden. Einige Datenbanken bieten auch eine sogenannte API an. 
Diese kann dazu genutzt werden, direkt ohne Umweg über ein Browser-Tab eine Suchanfrage zu stellen.

In diesem Beispiel ergibt die Suchanfrage einen Webseiten String, der einer Datenbank-Abfrage entspricht. Nun kann über das requests Paket eine Abfrage der Webseite gestartet werden und
das Resultat wird mittels BeautifulSoup in eine "Textsuppe" umgewandelt.

In [2]:
url = 'http://jordanus.badw.de/cgi-bin/lrz_jord-search.pl?sprache=en&datenbank=ptolemaeus&ausgabe=dhstext&ausdateiformat=txt&listpos=0&listen=keine&listlet=keiner&fn=t52a56.f&fi=Martianus+Capella'
    
r  = requests.get(url)

data = r.text

soup = BeautifulSoup(data,'lxml')

## Beautiful Soup Basics



Get the title of the html page

In [3]:
soup.title

<title>Search result</title>

Find all paragraphs

In [4]:
soup.findAll('p')

[<p>
 </p>]

Find all links

In [5]:
soup.findAll('a')[:4]

[<a href="http://jordanus.badw.de/cgi-bin/lrz_jord-form.pl?sprache=en"><font color="#8A2020" size="+1">New search</font>
 </a>,
 <a href="http://jordanus.badw.de/cgi-bin/lrz_jord-search.pl?sprache=en&amp;datenbank=ptolemaeus&amp;ausgabe=dhs&amp;ausdateiformat=txt&amp;listpos=0&amp;listen=keine&amp;listlet=keiner&amp;fn=t52a56.f&amp;fi=Martianus+Capella"><b>ACAPR00001/48</b></a>,
 <a href="http://jordanus.badw.de/cgi-bin/lrz_jord-search.pl?sprache=en&amp;datenbank=ptolemaeus&amp;ausgabe=dhs&amp;ausdateiformat=txt&amp;listpos=0&amp;listen=keine&amp;listlet=keiner&amp;fn=t52a56.f&amp;fi=Martianus+Capella"><b>ANEYHC3970726/02</b></a>,
 <a href="http://jordanus.badw.de/cgi-bin/lrz_jord-search.pl?sprache=en&amp;datenbank=ptolemaeus&amp;ausgabe=dhs&amp;ausdateiformat=txt&amp;listpos=0&amp;listen=keine&amp;listlet=keiner&amp;fn=t52a56.f&amp;fi=Martianus+Capella"><b>BBRGS0424/04</b></a>]

Die so erhaltenen Daten können nun je nach Struktur der Webseite nach Merkmalen durchsucht werden. Bei der Jordanus Webseite (http://jordanus.badw.de/) werden die Daten als Tabelle dargestellt. 
Deshalb wird nun nach dem Wort "table" gesucht. Jede Tabelle ist mit Zeilen ("tr") und Spalten ("td") strukturiert. Der Inhalt jeder Zeile wird in ein Schlüssel:Wert Paar umgewandelt und in eine Liste geschrieben.   

In [6]:
data = []
tables = soup.find_all('table')

for table in tables:
    item = {}
    table_body = table.find_all('tr')
    for row in table_body:
        cols = row.find_all('td')
        cols = [ele.text.strip() for ele in cols]
        if cols:
            item[cols[0]] = cols[1]
    data.append(item) 

In [7]:
data[:4]

[{'Search expr.:': 'Author = Martianus Capella'},
 {'Number of records found: \n82': 'New search'},
 {},
 {'2. Incipit': 'Item Machrobius. Luna item circuli',
  '2/48': 'ACAPR00001/48',
  'Add. numbers': 'Gen. Sci. 1',
  'Author standard': 'Macrobius\nMartianus Capella',
  'Century': '12\n13\n14\n15',
  'City': 'San Juan Capistrano',
  'Condition': '(152-154): these leaves are slit and rather crudely repaired. Apart from these three leaves the whole of this manuscript is in fine condition\nUse of a piece of parchment from an English language bill of conveyance to repair (150)',
  'Contents': 'Comm. 1.20.14 (p. 81, line 10) to 1.20.32 (p.84, line 27); extract from Felix Capella; (80r) Macrobius. Eratostenes philosophus ideoque geometra (...)',
  'Country': 'USA',
  'Cover': 'Bound in 15th century calf\nBound in old English calf',
  'Decorations': '(152v-154v) Colour drawings of constellations\n(152-154) contain 30 remarkable English pen-drawings of the early 13th century; initials mainl

## Erzeugung des Dataframes

Der Dataframe kann nun einfach erstellt werden. In einem ersten Schritt sind aber noch zu viele Schlüssel enthalten, die keine auswertbare Information ergeben. 

In [8]:
df = pd.DataFrame(data)
print(df.shape)
df.keys()

(85, 182)


Index(['10/3', '11/2', '12/3', '12/4', '13/2', '13/4', '13/5', '13/6', '14/3',
       '15/9',
       ...
       'Source', 'Source's source', 'Stype of writing', 'Subject 1',
       'Subject 2', 'T & K', 'Type of ms. 2', 'Type of source', 'Watermarks',
       'Width of columns'],
      dtype='object', length=182)

Um eine Untermenge des Dataframes auszuwählen, können wir reguläre Ausdrücke nutzen. Alle Schlüssel der Form "[Zahlen]/[Zahlen]" sollen nicht angezeigt werden. Einge neue Liste mit sinnvollen Schlüsseln wird erzeugt und als Auswahlkriterium an den Dataframe übergeben.  

In [9]:
keyList = [x for x in df.keys() if not re.findall('\d+/\d+',x)]

In [10]:
dfReducedCapella= df[keyList].dropna(axis=1,how='all')
dfReducedCapella.shape

(85, 75)

In [11]:
dfReducedCapella.head(5)

2. Incipit Add. numbers Addit. Names Addit. copies  \
0                                 NaN          NaN          NaN           NaN   
1                                 NaN          NaN          NaN           NaN   
2                                 NaN          NaN          NaN           NaN   
3  Item Machrobius. Luna item circuli  Gen. Sci. 1          NaN           NaN   
4                                 NaN          NaN          NaN           NaN   

  Addit. inform.             Addit. remarks Additional Texts Additional parts  \
0            NaN                        NaN              NaN              NaN   
1            NaN                        NaN              NaN              NaN   
2            NaN                        NaN              NaN              NaN   
3            NaN                        NaN              NaN              NaN   
4            NaN  Liber VIII: De astronomia              NaN              NaN   

                          Author standard Autor ms.  \
0                                     NaN       NaN   
1                                     NaN       NaN   
2                                     NaN       NaN   
3            Macrobius\nMartianus Capella       NaN   
4  Martianus Capella\nMacrobius (attrib.)       NaN   

                   ...                    \
0                  ...                     
1                  ...                     
2                  ...                     
3                  ...                     
4                  ...                     

                                              Source Source's source  \
0                                                NaN             NaN   
1                                                NaN             NaN   
2                                                NaN             NaN   
3  Faye/Bond (1962)\nAuktionskatalog Robinson\nN....             NaN   
4                                   Faulhaber (1983)             NaN   

                                    Stype of writing Subject 1 Subject 2  \
0                                                NaN       NaN       NaN   
1                                                NaN       NaN       NaN   
2                                                NaN       NaN       NaN   
3  (1-128) written in a very fine book-hand, the ...       NaN       NaN   
4                       Minute gothic minuscule hand       NaN       NaN   

     T & K Type of ms. 2                              Type of source  \
0      NaN           NaN                                         NaN   
1      NaN           NaN                                         NaN   
2      NaN           NaN                                         NaN   
3   TK 699           NaN  cat.\ncat.\nts. pvt. notes\nts. pvt. notes   
4  TK 1183           NaN                                        cat.   

                                          Watermarks  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  w : (inner ff.II., older than outer ones) crow...   

                        Width of columns  
0                                    NaN  
1                                    NaN  
2                                    NaN  
3                                    NaN  
4  166 x 86, first line below top ruling  

[5 rows x 75 columns]

Die gewünschten Schlüssel können auch direkt als Liste übergeben werden und so eine noch reduziertere Form des Dataframes erzeugen.

In [12]:
smallList = ['Author standard','Short title','Short title st.','Library','Language','Folio no.','City','Country','Century']
df[smallList].dropna(how='all').reset_index(drop=True).head(5)

Author standard                       Short title  \
0            Macrobius\nMartianus Capella                               NaN   
1  Martianus Capella\nMacrobius (attrib.)                               NaN   
2                       Martianus Capella                               NaN   
3                       Martianus Capella  Recueil de traités scientifiques   
4                       Martianus Capella                               NaN   

                                    Short title st.  \
0                 Selections from Macrobius (part.)   
1        De nuptiis Philologiae et Mercurii (part.)   
2                       Annotations de genre divers   
3                De nuptiis Philologiae et Mercurii   
4  De nuptiis Philologiae et Mercurii (part., inc.)   

                                      Library Language  Folio no.  \
0       The Library of Robert B. Honeyman Jr.    Latin    79r-80v   
1  Library of the Hispanic Society of America    Latin      1r-6v   
2             Stedelijke Openbare Bibliotheek    Latin  309r-311r   
3                         Bibliothèque Royale    Latin   13v-196v   
4                            Staatsbibliothek    Latin    65r-79r   

                  City  Country         Century  
0  San Juan Capistrano      USA  12\n13\n14\n15  
1             New York      USA             NaN  
2               Brugge  Belgium              14  
3            Bruxelles  Belgium              10  
4              Bamberg  Germany              10

## Suchanfrage II: isisCB explore

Die Webseite für das isisCB explore Tool der "Isis Bibliography of the History of Science" ist strukturell anders. Hier werden Suchergebnisse über eine Link-Liste dargestellt. 
In der Liste werden einzelne Ergebnisse in paragraphen sortiert (mit dem tag `<p>`) . Wir generieren eine Unterliste aller paragraphen die das Word `Article` oder `Book` enthalten. 

In [13]:
url2 = 'https://data.isiscb.org/isis/?q=copernicus&models=isisdata.citation&sort_order_citation=publication_date_for_sort&sort_order_dir_citation=descend&sort_order_dir_authority=ascend&selected_facets=citation_subject_ids_exact:CBA000021037'

r2  = requests.get(url2)

data2 = r2.text

soup2 = BeautifulSoup(data2,'lxml')

In [14]:
baseURL = 'https://data.isiscb.org/'

Suche alle Paragraphen, die im Text das Wort Article oder Book enthalten.

In [15]:
plist = soup2.findAll('p')
queListe = [x for x in plist if "Article" in x.text or "Book" in x.text]

Für alle gefundenen Datensätze müssen wir nun einem weiteren Link folgen. Dieser wird über den tag `<a>` gesucht, mit der Option `href=True`.
Mit dem requests Paket wird eine Anfrage `GET` an die URL gesendet, die sich aus der Basis-URL und dem gefundenen Link zusammensetzt. 

Das Resultat wird in eine Soup verwandelt. In der neuen Webseite werden die Informationen zu einem Artikel oder Buch über das Element `div` mit der Klasse `class=col-sm 5` dargestellt. 
In einem solchen Element stehen die Informationen wieder in Paragraphen, die in eine Untermenge eingetragen werden.

In [16]:
resultList = []
for x in queListe:
    links = x.findAll('a',href=True)
    for link in links:
        scndLevel = baseURL + link['href']
        scndRes = requests.get(scndLevel)
        dataTemp = scndRes.text
        soupTemp = BeautifulSoup(dataTemp,'lxml')
        x = soupTemp.findAll('div',class_='col-sm-5')
        pListTemp = x[0].findAll('p')
        subList = []
        for parag in pListTemp:
            res = parag.text
            subList.append(res)
        resultList.append(subList)    

Mittels regex wird aus der Menge aller Resultate eine Menge von Dictionaries gebaut, vorraus wir einen sortierten Dataframe erhalten.

In [17]:
dictList = []
for i in range(len(resultList)):
    subDict = {}
    for k in range(len(resultList[i])):
        keys = re.findall('.+?(?=:)',resultList[i][k])
        value =re.findall('(?<=:).+',resultList[i][k]) 
        if keys:
            if value:
                subDict[keys[0]] = value[0]#resultList[i][k]
    dictList.append(subDict)

In [18]:
dfISISTemp = pd.DataFrame(dictList)
dfISISTemp.head(5)

http  \
0  //data.isiscb.org/isis/citation/CBB001551118/   
1  //data.isiscb.org/isis/citation/CBB001551912/   
2  //data.isiscb.org/isis/citation/CBB001451084/   
3  //data.isiscb.org/isis/citation/CBB001510107/   
4  //data.isiscb.org/isis/citation/CBB001420393/   

    Adventures in the Bone Trade   Lee De Forest   Secrets of the Old One  \
0                            NaN             NaN                      NaN   
1                            NaN             NaN                      NaN   
2                            NaN             NaN                      NaN   
3                            NaN             NaN                      NaN   
4                            NaN             NaN                      NaN   

                                            Abstract  \
0   The contributors to Making of Copernicus exam...   
1   At the center of this article is an iconograp...   
2   Copernicus makes his dislike of the equant kn...   
3   Is science more rational or objective than an...   
4   During the two decades before the turning poi...   

                              Authors & Contributors Description  \
0   Neuber, Wolfgang (Author) ; Rahn, Thomas (Aut...         NaN   
1                         Gaulke, Karsten (Author) ;         NaN   
2                          Blåsjö, Viktor (Author) ;         NaN   
3                       Parsons, Keith M. (Author) ;         NaN   
4                    Blumenthal, Geoffrey (Author) ;         NaN   

  Publication Date  
0             2015  
1             2015  
2             2014  
3             2014  
4             2014

Um den Dataframe zu bereinigen, verwerfen wir alle Spalten in denen weniger als zwei Werte ungleich NaN sind. Zudem verwerfen wir alle Zeilen, die nur NaN Einträge enthalten.

In [19]:
dfISIS = dfISISTemp.dropna(axis=1,thresh=2)
dfISIS = dfISIS.dropna(axis=0,how='all')
dfISIS

http  \
0   //data.isiscb.org/isis/citation/CBB001551118/   
1   //data.isiscb.org/isis/citation/CBB001551912/   
2   //data.isiscb.org/isis/citation/CBB001451084/   
3   //data.isiscb.org/isis/citation/CBB001510107/   
4   //data.isiscb.org/isis/citation/CBB001420393/   
5   //data.isiscb.org/isis/citation/CBB001422065/   
6   //data.isiscb.org/isis/citation/CBB001321204/   
7   //data.isiscb.org/isis/citation/CBB001500389/   
8   //data.isiscb.org/isis/citation/CBB001510030/   
9   //data.isiscb.org/isis/citation/CBB001213931/   
10  //data.isiscb.org/isis/citation/CBB001213869/   
11  //data.isiscb.org/isis/citation/CBB001252571/   
12  //data.isiscb.org/isis/citation/CBB001212614/   
13  //data.isiscb.org/isis/citation/CBB001320811/   
14  //data.isiscb.org/isis/citation/CBB001200853/   
15  //data.isiscb.org/isis/citation/CBB001320808/   
16  //data.isiscb.org/isis/citation/CBB001252174/   
17  //data.isiscb.org/isis/citation/CBB001200855/   
18  //data.isiscb.org/isis/citation/CBB001250442/   

                                             Abstract  \
0    The contributors to Making of Copernicus exam...   
1    At the center of this article is an iconograp...   
2    Copernicus makes his dislike of the equant kn...   
3    Is science more rational or objective than an...   
4    During the two decades before the turning poi...   
5    With exoplanets being discovered daily, Earth...   
6    This essay studies Moses Galeano, a Jewish sc...   
7    William Shakespeare lived at a remarkable tim...   
8    This volume brings John Milton's Paradise Los...   
9    In Copernicus in the Cultural Debates of the ...   
10   The period that separates comproehendunt from...   
11                                                NaN   
12   Im Jahre 1618 editierte und publizierte Joann...   
13   The Ptolemy-Copernicus transition is analyzed...   
14                                                NaN   
15   In 1454 Georg Peurbach teaches astronomy at t...   
16   The Copernican Question advances a radical re...   
17                                                NaN   
18                                                NaN   

                               Authors & Contributors Publication Date  
0    Neuber, Wolfgang (Author) ; Rahn, Thomas (Aut...             2015  
1                          Gaulke, Karsten (Author) ;             2015  
2                           Blåsjö, Viktor (Author) ;             2014  
3                        Parsons, Keith M. (Author) ;             2014  
4                     Blumenthal, Geoffrey (Author) ;             2014  
5                          Gingerich, Owen (Author) ;             2014  
6                         Morrison, Robert (Author) ;             2014  
7                                Falk, Dan (Author) ;             2014  
8                Danielson, Dennis Richard (Author) ;             2014  
9                    Omodeo, Pietro Daniel (Author) ;             2014  
10   Segonds, Alain-Philippe (Author) ; Luna, Conc...             2013  
11                    Blumenthal, Geoffrey (Author) ;             2013  
12                         Chroboczek, Jan (Author) ;             2013  
13                       Nugayev, Rinat M. (Author) ;             2013  
14                           Barker, Peter (Author) ;             2013  
15                    Malpangotto, Michela (Author) ;             2013  
16                      Westman, Robert S. (Author) ;             2013  
17                  Graney, Christopher M. (Author) ;             2013  
18                           Richards, Ted (Author) ;             2012

## Suchanfrage III: Newtons Gesammelte Werke 

Die Homepage des Newton Projects der University of Sussex, http://www.newtonproject.sussex.ac.uk/, bietet besser strukturierte Inhalte. 
Hier können die Informationen mit passenden Klassennamen sinnvoll ausgelesen werden. 

Die genauen Spezifikation sind als PDF verfügbar: http://www.newtonproject.sussex.ac.uk/resources/pdfs/techspec.pdf

In [3]:
url3 = 'http://www.newtonproject.sussex.ac.uk/prism.php?id=43'

r3  = requests.get(url3)

data3 = r3.text

soup3 = BeautifulSoup(data3,'lxml')

In [4]:
baseURLNewton = 'http://www.newtonproject.sussex.ac.uk'

Die Einträge werden in einer Table mit der Klasse `record` ausgegeben. 

In [5]:
allrecords = soup3.findAll('td',class_='record')

Der Title eines Eintrags:

In [6]:
allrecords[0].findAll('p',class_='title')[0].text

"Fair copies of the 'Short Chronicle' and 'Chronology of Ancient Kingdoms Amended'"

den Author erhalten wir über 

In [7]:
allrecords[0].findAll('p',class_='author')[0].text

'Author: Isaac Newton'

ebenso verfahren wir mit metadata, source, links und primary_key. Wieder wird eine Liste mit Sub-Dictionaries erzeugt und daraus ein Dataframe gebaut. 

Um den normalisierten Volltext zu erhalten, folgen wir wieder einem Link auf die Unterseite und extrahieren dort den Text. Dieser ist eindeutig markiert mit der id `tei`
Über den Befehl `.text` erhalten wir den Text ohne die HTML-Struktur.

**Achtung:**

Das Abfragen von allen Einträgen dauert etwas länger. Daher werden unten nur die ersten 10 Einträge abgefragt (`[:10]`) um den Server der Uni Sussex nicht zu überlasten. Wenn die Texte mit NLTK weiter bearbeitet werden sollen, bietet es sich daher an, den resultierenden Dataframe als Pickle abzuspeichern. 

In [332]:
dictListNewton = []
for record in soup3.findAll('td',class_='record')[:10]:
    subDict = {}
    for keyClass in ['title', 'author','metadata','source','primary_key']:
        try:
            value = record.findAll('p',class_=keyClass)[0].text
            subDict[keyClass] = value
        except:
            subDict[keyClass] = None
    links = record.findAll('a',href=True,text='Normalized\xa0Text')
    if links:
        scndLevel = baseURLNewton + links[0]['href']
        scndRes = requests.get(scndLevel)
        dataTemp = scndRes.text
        soupTemp = BeautifulSoup(dataTemp,'lxml')
        text = soupTemp.findAll('div',id='tei')
        subDict['norm_text'] = text[0]
    dictListNewton.append(subDict)    

In [333]:
dfNewton = pd.DataFrame(dictListNewton)
dfNewton.head(3)

author                                           metadata  \
0  Author: Isaac Newton     Metadata: in English, c. 86,051 words, 120 ff.   
1  Author: Isaac Newton  Metadata: in Latin, c. 6,207 words, 21 pp. on ...   
2  Author: Isaac Newton  Metadata: mainly in Latin with some English, G...   

                                           norm_text  \
0  <div id="tei">
<div>
<span class="pagenumber" ...   
1  <div id="tei">
<div>
<span class="pagenumber" ...   
2  <div id="tei">
<div>
<span class="pagenumber" ...   

                      primary_key  \
0  Newton Catalogue ID: THEM00090   
1  Newton Catalogue ID: THEM00046   
2  Newton Catalogue ID: THEM00056   

                                              source  \
0  Source: MS Add. 3988, Cambridge University Lib...   
1  Source: Yahuda Ms. 3, National Library of Isra...   
2  Source: Yahuda Ms. 13, National Library of Isr...   

                                               title  
0  Fair copies of the 'Short Chronicle' and 'Chro...  
1  'Introductio. Continens Apocalypseos rationem ...  
2       Miscellaneous theological extracts and notes

In [334]:
dfNewton.shape

(250, 6)

### Speichern und Laden von Dataframes

Die Struktur von Dataframes bietet es an, diese als JSON zu speichern. Das erfolgt einfach mit df.to_json('Dateiname'). Über die normale Einleseroutine von JSONs kann dann der Dataframe aus der Datei wieder hergestellt werden. Die bereits vorhandene JSON Datei enthält alle 250 Einträge zu Newton.

In [335]:
# Save dataframe to json file on disk
#dfNewton.to_json('./newton_metadata_frame.json')

In [329]:
# Load by using
# with open('./newton_metadata_frame.json') as NewtonData:
#     newtonJson = json.load(NewtonData)
#     dfTest = pd.DataFrame(newtonJson)

### Weitere Struktur: Paragraphen je Kapitel je Text

Um eine weitergehende Textanalyse durchzuführen, bietet es sich an, die Texte nach Paragraphen zu strukturieren und dann darauf Anlysemethoden anzuwenden.

In [336]:
chpStructure = []
for i in range(len(dfNewton)):
    title = str(i) + ': ' +  dfNewton['title'].iloc[i]
    try:
        #Konstruiere Liste aller Kapitel
        chps = dfNewton['norm_text'].iloc[i].findAll('div')
    except:
        pass
    if chps:
        for chp in range(len(chps)):
            pgsTemp = {}
            chKey = str(chp)
            # Konstruiere Liste aller Absätze
            paras = chps[chp].findAll('p')
            for pg in range(len(paras)):
                #print('Chp:{0},Para:{1}'.format(chp,pg))
                keyTemp = str(pg)
                valueTemp = paras[pg].text
                # Speichere den Text des Paragraphen mit Buch,Kapitel, und Paragraphennummer als tuple
                triple = (title,chp,pg,valueTemp)
                chpStructure.append(triple)

In [345]:
dfNewtonParagraphs = pd.DataFrame.from_dict(chpStructure)
dfNewtonParagraphs.set_index(0, inplace=True) # Setze Text-Titel als Index
dfNewtonParagraphs.set_index(1, append=True, inplace=True) # Ergänze Kaptitel als zweite Ebene Index
dfNewtonParagraphs.index.rename(['title','chapter'], inplace=True)
dfNewtonParagraphs.columns = ['paragraph','paragraph_text'] # Benenne Spalten um
print('Anzahl der Paragraphen aller Newton Texte: {}'.format(dfNewtonParagraphs.shape[0]))
dfNewtonParagraphs.head()

Anzahl der Paragraphen aller Newton Texte: 33170


paragraph  \
title                                              chapter              
0: Fair copies of the 'Short Chronicle' and 'Ch... 0                0   
                                                   0                1   
                                                   1                0   
                                                   1                1   
                                                   1                2   

                                                                                               paragraph_text  
title                                              chapter                                                     
0: Fair copies of the 'Short Chronicle' and 'Ch... 0        This is the original manuscript from which Sir...  
                                                   0                                           Iohn Conduitt.  
                                                   1        The Greek Antiquities are full of poetical fic...  
                                                   1        So then a little after the death of Alexander ...  
                                                   1        But how uncertain their Chronology is, & how d...

In [351]:
#dfNewtonParagraphs.reset_index().to_json('./newton_paragraphs_frame.json')

In [349]:
# Load by using
#with open('./newton_paragraphs_frame.json') as NewtonParaData:
#    newtonParaJson = json.load(NewtonParaData)
#    dfTestPara = pd.DataFrame(newtonParaJson)
#dfTestPara

# Aus externeren JSON Daten

Auch mit Dritt-Programmen erstellte JSON Dateien können die Grundlage für eine Websuche bilden. Eine Liste mit 1000 Publikation zur Kalten Fusion dient hier als Beispiel.

In [6]:
with open('./PoPCites.json','r',encoding='utf-8-sig') as coldFusionData:
    coldFusionJson = json.load(coldFusionData)

In [7]:
dfColdFusion = pd.DataFrame(coldFusionJson)

In [8]:
dfColdFusion.keys()

Index(['article_url', 'authors', 'cites', 'cites_url', 'ecc', 'publisher',
       'rank', 'source', 'title', 'type', 'use', 'year'],
      dtype='object')

In [21]:
dfColdFusion.sort_values(['year'],ascending=False).head(10)

article_url  \
500               http://www.jstor.org/stable/42577264   
812                                                NaN   
391  https://inis.iaea.org/search/search.aspx?orig_...   
816  http://en.cnki.com.cn/Article_en/CJFDTOTAL-HJB...   
393  http://www.tandfonline.com/doi/abs/10.1080/188...   
394       http://www.ans.org/pubs/journals/nse/a_27466   
814  http://oai.dtic.mil/oai/oai?verb=getRecord&met...   
813  http://www.tandfonline.com/doi/abs/10.1080/009...   
399  http://link.springer.com/article/10.1007/BF015...   
810  http://journals.aps.org/prc/abstract/10.1103/P...   

                                               authors  cites  \
500                                       [R Wanderer]    NaN   
812      [LL Zahm, AC Klein, SE Binney, JN Reyes, ...]   12.0   
391  [Z Wilhelmi, Z Szeflinski, J Tarasiuk, A Turow...    1.0   
816                                   [QINJWLY Dingpu]    3.0   
393                           [I Yamamoto, A Kanagawa]   47.0   
394          [FH Coensgen, TA Casper, DL Correll, ...]   26.0   
814  [DR Lawson, MJ Tierney, IF Cheng, LS Van Dyke,...    NaN   
813                                     [JG Patterson]    1.0   
399                [GS Collins, JS Walker, JW Norbury]    1.0   
810                            [SE Koonin, M Mukerjee]   19.0   

                                             cites_url   ecc  \
500                                                NaN   NaN   
812  https://scholar.google.com/scholar?cites=70272...  12.0   
391  https://scholar.google.com/scholar?cites=37188...   1.0   
816  https://scholar.google.com/scholar?cites=43233...   3.0   
393  https://scholar.google.com/scholar?cites=11976...  47.0   
394  https://scholar.google.com/scholar?cites=17170...  26.0   
814                                                NaN   NaN   
813  https://scholar.google.com/scholar?cites=15270...   1.0   
399  https://scholar.google.com/scholar?cites=16609...   1.0   
810  https://scholar.google.com/scholar?cites=14787...  19.0   

            publisher  rank                               source  \
500             JSTOR   501   ETC: A Review of General Semantics   
812          Elsevier   813        … chemistry and interfacial …   
391     inis.iaea.org   392                           Nukleonika   
816    en.cnki.com.cn   817    Nuclear Fusion and Plasma Physics   
393  Taylor & Francis   394  … of Nuclear Science and Technology   
394           ans.org   395                Nuclear Science and …   
814     DTIC Document   815                                  NaN   
813  Taylor & Francis   814                       Energy sources   
399          Springer   400             Journal of Fusion Energy   
810               APS   811                    Physical Review C   

                                                 title      type   use  year  
500  Cold Fusion: A Profusion of Illusion, Confusio...       NaN  True  1990  
812  Experimental investigations of the electrolysi...  CITATION  True  1990  
391  Search for neutron emission in the deuterium-p...       NaN  True  1990  
816  HYSTERESIS BIFURCATION AND MULTILEVEL JUMP IN ...  CITATION  True  1990  
393  Possibility of remarkable enhancement of separ...       NaN  True  1990  
394  High-Performance Beam-Plasma neutron sources f...       NaN  True  1990  
814  Coulometric Assay of Deuterium Loading Levels ...       NaN  True  1990  
813  Global change in carbon combustion: trends in ...       NaN  True  1990  
399       Deuteron tunneling at electron-volt energies       NaN  True  1990  
810  Branching ratios in low-energy deuteron-induce...       NaN  True  1990

In [23]:
dfColdFusion['article_url'].iloc[393]

'http://www.tandfonline.com/doi/abs/10.1080/18811248.1990.9731176'

Die meisten Artikel verfügen über eine URL Angabe, die aber teilweise auf nicht mehr funktionierende Seiten führt. 

Große Verlage, wie zB. Physical Review Letters, blockieren zudem die automatisierte PDF-Suche durch Captchas. 
Für wissenschaftliche Arbeiten kann teilweise ein extra Zugang zu Daten beantragt werden.

Ist das PDF per Hand herunter geladen, kann mittels PyPDF2 der Text extrahiert werden. 
Je nach Verarbeitung des PDFs sind allerdings die resultierenden Strukturen sehr unterschiedlich.

In unterem Beispiel wird die erste Seite ausgenommen, da sie Metadaten des Download-Zeitpunktes enthält und keinen extrahierbaren Text.

In [36]:
import PyPDF2

pdfText = []

with open('./Possibility of Remarkable Enhancement of Separation Factor by Cryogenic Wall Thermal Diffusion Column.pdf','rb') as pdfDaten:
    pdfReader = PyPDF2.PdfFileReader(pdfDaten)
    print('Number of pages: {}'.format(pdfReader.numPages))
    for num in range(1,pdfReader.numPages):
        pageObj = pdfReader.getPage(num)
        ts=pageObj.extractText()
        pdfText.append(ts)
pdfText[0]

Number of pages: 7


'Journal of NUCLEAR SCIENCE and TECHNOLOGY, 27[3), pp. 250-255 (March 1990). Possibility of Remarkable Enhancement of Separation Factor by "Cryogenic-Wall" Thermal Diffusion Column Ichiro YAMAMOTO and Akira KANAGA W A \nDepartment of Nuclear Engineering, Faculty of Engineering, Nagoya University* Received july 3, 1989 From a view point that the maximum separation factor (af3lmax attainable in total reflux operation of thermal diffusion column depends on temperature difference JT between hot Th and cold Tc surfaces in the form of JT/Tc as (a ) =ex [_&7\'_ ___ __ /{l-E_· 4+3ln_o_·}] (3 max P Tc 2lnilv\'15(lno)2+56lnil+53 Tc 4lno(l+lno) , where il C==rhfrc) is the ratio of hot wire to cold-wall radii, ar the thermal diffusion factor and zŁ (==.Zfrc) the normalized column height, remarkable enhancement of separation factor is expected through an adoption of "Cryogenic-Wall" (ex. 77.35 K) instead of "Ordinary wall" (ex. Tc=288.15 K). The effect of "Cryogenic-Wall" was analyzed on the basis 

Wenn die Struktur der einzelnen Seiten nicht von Bedeutung ist, können alle Seiten zusammengefasst werden und dann Abschnitte mittels Regex gesucht werden.

In [38]:
pdfString = ' '.join(pdfText)

Authoren haben in diesem Journals das Format Vorname NACHNAME. Mittels regex können wir also die Authoren extrahieren.

In [57]:
re.findall('[A-Z][a-z]{3,10} [A-Z]{3,10}',pdfString)

['Ichiro YAMAMOTO', 'Akira KANAGA']

Das bedeutet aber auch, dass die Abschnitte des PDFs nicht die einzigen Wörter mit Großbuchstaben sind. Insbesondere sind die Referenzen ein Problem. Daher wird der PDF string in drei Abschnitte geteilt: Header, Haupttext und Referenzen. 

In [69]:
refs = pdfString.split('REFERENCES')[1]
tmp = pdfString.split('REFERENCES')[0]
head = tmp.split('KEYWORDS')[0]
main = tmp.split('KEYWORDS')[1]

Nun können alle Abschnitt-Überschriften des Haupttextes gefunden werden. 

In [102]:
capitalWords = re.findall('[A-Z]{5,15}',main)

In [103]:
capitalWords

['INTRODUCTION',
 'OPTIMUM',
 'PRESSURE',
 'MAXIMUM',
 'SEPARATION',
 'FACTOR',
 'RESULTS',
 'DISCUSSION',
 'UIOCI',
 'CONCLUSION',
 'ACKNOWLEDGMENT']

Um die einzelnen Abschnitte automatisiert zu erhalten, ergänzen wir die Liste mit der Überschrift 'References' und extrahieren dann mit regex den string zwischen zwei Überschriften. 
Ist dieser Text länger als 10 Zeichen, hängen wir das Ergebnis an die sections Liste an. 

In [109]:
sections = []

capitalWords.append('REFERENCES')

for i in range(len(capitalWords)-1): 
    sW = capitalWords[i]
    eW = capitalWords[i+1]
    #pat = re.compile('(?<=INTRODUCTION).+(?=OPTIMUM PRESSURE)',flags=re.DOTALL)
    pat = re.compile('(?<={0}).+(?={1})'.format(sW,eW),flags=re.DOTALL)
    res = re.findall(pat,pdfString)
    if res and len(res[0]) > 10:
        sections.append(res[0])

In [116]:
sections[0]

' Thermal diffusion<ll-<Ł> is not suitable for large scale production because of low dynamic efficiency. Gas phase thermal sion of hydrogen is, however, a convenient method for the separation of the isotopes of hydrogen on a small scale to moderate scale<Ł>, because the process has advantages such as small tritium inventory, high separation factors in apparatus of moderate size and a relatively simple system. The process, therefore, attracts attention to purification of tnttum in tion system and recovery of tritium from used gas mixture of hydrogen isotopes in fusion fuel cycle. At the Tritium Process Laboratory (TPL) in the Japan Atomic Energy Research Institute, the thermal diffusion process is plied for recovery and enrichment of tritium, which is diluted and mixed with other drogen isotopes in the other experiments<Ł>. In previous papers<n-<Ł>, we have reported results for H2-HT isotope separation; (1) an * Furo-cho, Chikusa-ku, Nagoya 46 4-01. -58- Vol. 27, No. 3 (Mar. 1990) exper

Bis auf einen Abschnitt (2 und 3 sind eigentlich ein Abschnitt) werden alle Einheiten gut erkannt. 

# Geographische Daten

Antike Orte in dem Projekt https://pleiades.stoa.org/ können über eine API abgefragt werden: http://api.pleiades.stoa.org/

Zeige den aktuellen Zustand des Projekts

In [418]:
res = requests.get('http://api.pleiades.stoa.org/status')
res.json()

{'num_locations': 39363, 'num_names': 31012, 'num_places': 35204}

Erhalte Informationen zu einer bestimmten Stadt, kodiert durch ID.

In [419]:
res2 = requests.get('http://pleiades.stoa.org/places/423025/json')
jsonRes = res2.json()

In [420]:
jsonRes.keys()

dict_keys(['names', 'features', 'uri', '@type', 'id', 'description', 'rights', 'type', 'references', 'details', 'provenance', 'title', 'placeTypes', 'created', '@context', 'contributors', 'subject', 'bbox', 'creators', 'review_state', 'history', 'connectsWith', 'locations', 'reprPoint'])

In [421]:
dfRoma = pd.DataFrame([jsonRes])
dfRoma

@context  @type  \
0  {'created': 'dcterms:created', 'subject': 'dct...  Place   

                                       bbox  \
0  [12.486137, 41.8917, 12.4862, 41.891775]   

                                        connectsWith  \
0  [https://pleiades.stoa.org/places/303999556, h...   

                                        contributors               created  \
0  [{'username': None, 'name': 'DARMC'}, {'userna...  2010-11-10T22:33:36Z   

                                            creators  \
0  [{'username': None, 'name': 'L. Quilici'}, {'u...   

                                     description  \
0  The capital of the Roman Republic and Empire.   

                                             details  \
0  <p>The Barrington Atlas Directory notes: Roma/...   

                                            features  \
0  [{'properties': {'location_precision': 'precis...   

                    ...                                        placeTypes  \
0                   ...                     [urban, settlement, temple-2]   

                            provenance  \
0  Barrington Atlas: BAtlas 43 B2 Roma   

                                          references  \
0  [{'type': 'seeFurther', 'otherIdentifier': ' '...   

                          reprPoint review_state  \
0  [12.4861685, 41.891737500000005]    published   

                                              rights  \
0  Copyright © The Contributors. Sharing and remi...   

                                             subject title               type  \
0  [dare:major=1, dare:feature=major settlement, ...  Roma  FeatureCollection   

                                       uri  
0  https://pleiades.stoa.org/places/423025  

[1 rows x 24 columns]

Zeige verknüpfte Straßen zu dem Hauptort. Hierzu wird für jede Adresse in der Liste `dfRoma[connectsWith]` eine Anfrage gestellt und aus dem resultierenden JSON der Titel ausgelesen. 

In diesem Beispiel werden nur die ersten 20 Links abgefragt. 

In [422]:
connectedStreets = []
for i in (dfRoma['connectsWith'].iloc[0])[:20]:
    res = requests.get(i + '/json')
    try:
        jsonRes = res.json()
        title = jsonRes['title']
        connectedStreets.append(title)
    except:
        pass

In [423]:
connectedStreets

['Via Portuensis?',
 'Via Aemilia Scauri',
 'Via Latina',
 'Via Cornelia',
 'Via Nomentana',
 'Via Ostiensis',
 'Via Ardeatina',
 'Via Tiburtina',
 'Via Collatina',
 'Via Appia',
 'Via Clodia',
 'Via Aurelia',
 'Via Cassia',
 'Via Salaria',
 'Via Flaminia',
 'Via Praenestina',
 'Tiberis (river)',
 'Via Triumphalis',
 'Via Sublacensis']